### PR TITLE
add unit tests for project component

### PR DESCRIPTION
**Pull request by seanknox**
~~This is a WIP but hoping for some feedback and help.~~

First addition of unit tests for the project component. This is tricky as we use `ActivatedRoute` to feed data to the component via a route resolver; it's not a direct call to a service. We do this so data is pre-fetched before loading the page. 

Moreover, it seems as if the prefetch is not happening in the spec—see the introduction of the `?` throughout the template just to get the happy path tests to pass.

- [x] adds unit tests for project component
- [x] introduces factories for quickly creating mock data with https://github.com/willryan/factory.ts

### CODE REVIEW

File: src/app/project/project.component.spec.ts [hunk 1]
```diff
@@ -59,12 +36,137 @@ describe('ProjectComponent', () => {
...
+
+  describe('when project and build data is unavailable', () => {
+    beforeEach(() => {
+      project = undefined;
+      builds = undefined;
+    });
+
+    it('should create the component', () => {
+      expect(component).toBeTruthy();
+    });
```
Comment (seanknox) [inline src/app/project/project.component.spec.ts]: Interestingly (to me) this expectation must be run before in any describe block that tests component properties.

File: src/app/project/project.component.spec.ts [hunk 2]
```diff
@@ -3,38 +3,18 @@ import { RouterTestingModule } from '@angular/router/testing';
...
-    },
-    secrets: {},
-    worker: {
-      registry: '',
-      name: '',
-      tag: '',
-      pullPolicy: ''
-    }
-  };
+  let builds;
```
Comment (radu-matei) [inline src/app/project/project.component.spec.ts]: Can we also have the test strongly-typed?
(Not a blocker for this PR though)

File: src/app/tests/build-factory.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import * as Factory from 'factory.ts';
+
+import { Build } from '../models/Build';
+
+export const BuildFactory = Factory.makeFactory<Build>({
```
Comment (radu-matei) [inline src/app/tests/build-factory.ts]: Nice, really like this 😄

File: src/app/project/project.component.spec.ts (hunk 2 above)
Comment (seanknox) [reply src/app/project/project.component.spec.ts]: Done, thanks for catching.

**Comment by seanknox** [comment]
thanks @radu-matei, made changes you suggested. @technosophos I updated the spec to test the component DOM elements, so this is ready for your review.

File: src/app/project/project.component.html [hunk 4]
```diff
@@ -4,16 +4,16 @@
     <div class="breadcrumb-bar columns">
       <a routerLink="/">Home</a>
       <i class="icon ion-ios-arrow-forward"></i>
-      <a routerLink="project/:id">Project ID {{ project.id }}</a>
+      <a routerLink="project/:id">Project ID {{ project?.id }}</a>
```
Comment (seanknox) [inline src/app/project/project.component.html]: For some reason the unit tests don't seem to prefetch data like the real app, so I had to add these.

**Pull request merged**